Initial commit for new test framework of CI/CD

rocknes opened the pull request:
This checkin has 
1. the initial test framework implementation
2. ported tests in vcd_disk.py that can run in this new framework

Known issue : warning for insecure requests are not disabled

Sample test run,
(pyvcloud) C:\code\pyvcloud\system_tests>python -m unittest idisk_tests.TestDisk -v
vc1 is already attached.
Defaulting to first pvdc in the system viz.pvdc-2018-03-28-00-48-07.639
Reusing existing org pyvcloud-stystem-test-org.
Reusing existing user catalog_author.
Reusing existing user console_user.
Reusing existing user org_admin.
Reusing existing user vapp_author.
Reusing existing user vapp_user.
Reusing existing ovdc pyvcloud-system-test-org-vdc.
Reusing existing catalog test-cat
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:27: ResourceWarning: unclosed <ssl.SSLSocket fd=984, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54288), raddr=('10.150.198.193', 443)>
  Environment.create_catalog()
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:27: ResourceWarning: unclosed <ssl.SSLSocket fd=968, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54287), raddr=('10.150.198.193', 443)>
  Environment.create_catalog()
Reusing existing template test_vapp_template.ova
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:29: ResourceWarning: unclosed <ssl.SSLSocket fd=472, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54290), raddr=('10.150.198.193', 443)>
  Environment.upload_template()
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:29: ResourceWarning: unclosed <ssl.SSLSocket fd=552, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54289), raddr=('10.150.198.193', 443)>
  Environment.upload_template()
Reusing existing vApp test_vapp.
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:30: ResourceWarning: unclosed <ssl.SSLSocket fd=1008, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54292), raddr=('10.150.198.193', 443)>
  Environment.instantiate_vapp()
c:\code\pyvcloud\pyvcloud\system_test_framework\base_test.py:30: ResourceWarning: unclosed <ssl.SSLSocket fd=960, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54291), raddr=('10.150.198.193', 443)>
  Environment.instantiate_vapp()
test_0000_setup (idisk_tests.TestDisk) ... Creating disk : test_idisk
Created disk with id:38cbf816-c28c-4eb2-a9ab-f400cdc718cc
Creating disk : test_idisk_2
Created disk with id:879761ad-07e4-4791-9964-9710a7103795
Creating disk : test_idisk_2
Created disk with id:72af07f5-0c07-4717-8d80-6d595f98b947
ok
test_0010_get_all_disks (idisk_tests.TestDisk) ... ok
test_0020_get_disk_by_name (idisk_tests.TestDisk) ... ok
test_0030_get_disk_by_id (idisk_tests.TestDisk) ... ok
test_0040_change_idisk_owner (idisk_tests.TestDisk) ... c:\users\sena\appdata\local\programs\python\python36-32\Lib\unittest\case.py:605: ResourceWarning: unclosed <ssl.SSLSocket fd=372, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54300), raddr=('10.150.198.193', 443)>
  testMethod()
c:\users\sena\appdata\local\programs\python\python36-32\Lib\unittest\case.py:605: ResourceWarning: unclosed <ssl.SSLSocket fd=188, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54299), raddr=('10.150.198.193', 443)>
  testMethod()
ok
test_0050_attach_disk_to_vm_in_vapp (idisk_tests.TestDisk) ... ok
test_0060_detach_disk_from_vm_in_vapp (idisk_tests.TestDisk) ... c:\users\sena\appdata\local\programs\python\python36-32\Lib\email\feedparser.py:89: ResourceWarning: unclosed <ssl.SSLSocket fd=968, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54301), raddr=('10.150.198.193', 443)>
  for ateof in reversed(self._eofstack):
c:\users\sena\appdata\local\programs\python\python36-32\Lib\email\feedparser.py:89: ResourceWarning: unclosed <ssl.SSLSocket fd=188, family=AddressFamily.AF_INET, type=SocketKind.SOCK_STREAM, proto=0, laddr=('10.32.50.128', 54302), raddr=('10.150.198.193', 443)>
  for ateof in reversed(self._eofstack):
ok
test_0070_update_disk (idisk_tests.TestDisk) ... ok
test_9999_teardown (idisk_tests.TestDisk) ... ok

----------------------------------------------------------------------
Ran 9 tests in 136.366s

OK

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vmware/pyvcloud/227)
<!-- Reviewable:end -->

sahithi commented:
Reviewed 1 of 5 files at r1.
Review status: 1 of 5 files reviewed at latest revision, all discussions resolved.

---

*[neo_tests/base_config.yaml, line 4 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-6SPr87B_H2my-v2D:-LA-6SPr87B_H2my-v2E:b1e16fr) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L4)):*
> ```YAML
>   developer_mode: False
> 
> rest:
> ```

What is rest?
Can we have a better name?

---

*[neo_tests/base_test.py, line 15 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-Q_VbD3dGwgr5XZAo:-LA-Q_VbD3dGwgr5XZAp:b-b1miw2) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L15)):*
> ```Python
>     @classmethod
>     def setUpClass(cls):
>         if 'VCD_TEST_BASE_CONFIG_FILE' in os.environ:
> ```

Does it mean we cannot run tests with config.yaml passed as command-line param?

---

*[neo_tests/environment.py, line 70 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-mBRl7OAWmb4NHW6w:-LA-mBRl7OAWmb4NHW6x:bdyi72r) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L70)):*
> ```Python
>             raise Exception('Missing base configuration.')
>         if cls._sys_admin_client is None:
>             cls._sys_admin_client = cls.get_client(CommonRoles.SYS_ADMIN)
> ```

What is this cls? Who is passing in this param?
Where is  cls.get_client() and few other methods defined?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

sahithi commented:
Partial only.

---

Review status: 1 of 5 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-:-LA-mZKb-FCCWBQswYow:b-fmj2oy)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Review status: 1 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[neo_tests/base_config.yaml, line 4 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-6SPr87B_H2my-v2D:-LA-r2ssEu-E_AGFgYyk:b-e3fhiw) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L4)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

What is rest?
Can we have a better name?
</blockquote></details>

These configs are related to the rest calls we are making.  Right now it's not working as designed, since we are still getting warnings, I can rename t to something else. Any suggestions?

---

*[neo_tests/base_test.py, line 15 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-Q_VbD3dGwgr5XZAo:-LA-qWx5E22bKgn_Ui3Y:b-cdf55k) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L15)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Does it mean we cannot run tests with config.yaml passed as command-line param?
</blockquote></details>

As long as a config file with the default name is present in the test folder, we will pick it up and be able to run the test, we don't need to pass one explicitly while runnign the test.

_config_file = 'base_config.yaml'

The environment variable is to override the deafult config file name.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

sahithi commented:
Reviewed 4 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[neo_tests/environment.py, line 2 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-oIU_5Ci4nYeKv5ay:-LA-oIU_5Ci4nYeKv5az:b-i9ch3h) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L2)):*
> ```Python
> import warnings
> ```

Add docstrings for all the methods.

---

*[neo_tests/environment.py, line 109 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-ntuQBO7ZFTuxcJfk:-LA-ntuQBO7ZFTuxcJfl:b-2qe3o7) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L109)):*
> ```Python
>         vc_name = cls._config['vc']['vcenter_host_name']
>         for record in platform.list_vcenters():
>             if record.get('name') == vc_name:
> ```

What happens if listed VC name is not found? Throw an exception with right message here.

---

*[neo_tests/environment.py, line 120 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-ncVQ4W1dz_UAq1az:-LA-ncVQ4W1dz_UAq1b-:ba3fnih) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L120)):*
<details><summary><i alt="summary">Quoted 11 lines of code…</i></summary>

> ```
> # Untested code
>         platform.attach_vcenter(
>             vc_server_name=cls._config['vc']['vcenter_host_name'],
>             vc_server_host=cls._config['vc']['vcenter_host_ip'],
>             vc_admin_user=cls._config['vc']['vcenter_admin_username'],
>             vc_admin_pwd=cls._config['vc']['vcenter_admin_password'],
>             nsx_server_name=cls._config['nsx']['nsx_hostname'],
>             nsx_host=cls._config['nsx']['nsx_host_ip'],
>             nsx_admin_user=cls._config['nsx']['nsx_admin_username'],
>             nsx_admin_pwd=cls._config['nsx']['nsx_admin_password'],
>             is_enabled=True)
> ```
</details>


Please file a Jira task to track this and mention the taskId in TODO here

---

*[neo_tests/environment.py, line 126 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-o_cY64I7Gijl54Q_:-LA-o_cY64I7Gijl54Qa:b6pj1g5) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L126)):*
> ```Python
> 
>     @classmethod
>     def create_pvdc(cls):
> ```

This method seems to combine the logic of getting and existing pVdc and creation of new Pvdc (which is yet to be implemented).
1. Separate the logic into getExistingPvdc() and createPvdc()
2. Please file Jira task for TODO item - I suggest file one Umbrella task for all pending items in the framework and list all the TODO items there. Mention the Jira task here.

---

*[neo_tests/environment.py, line 136 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-pJp0BdbVrwIwQJcq:-LA-pJp0BdbVrwIwQJcr:b-tghmi3) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L136)):*
> ```Python
>         if pvdc_name is not '*':
>             for pvdc_ref in pvdc_refs:
>                 if pvdc_ref.get('name') == pvdc_name:
> ```

Add an else block to throw right exception here.

---

*[neo_tests/environment.py, line 137 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-p9FY1vmH4VTI24T6:-LA-p9FY1vmH4VTI24T7:bltxj2v) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L137)):*
> ```Python
>             for pvdc_ref in pvdc_refs:
>                 if pvdc_ref.get('name') == pvdc_name:
>                     print('Resusing existing ' + pvdc_name)
> ```

nit: reusing

---

*[neo_tests/environment.py, line 160 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qTDz-EK96lcUVMgZ:-LA-qTDz-EK96lcUVMg_:b-irw156) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L160)):*
> ```Python
>                 print('Reusing existing org ' + org_name + '.')
>                 cls._org_href = org.get('href')
>                 print('Using ' + cls._org_href)
> ```

consider clubbing above two print statements

---

*[neo_tests/environment.py, line 162 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qHTfA5SZDwr0geXF:-LA-qHTfA5SZDwr0geXG:b-5qo5dj) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L162)):*
> ```Python
>                 print('Using ' + cls._org_href)
>                 return
>         print('Creating new org ' + org_name)
> ```

Can we use logger class that Paco created recently instead of print?

---

*[neo_tests/environment.py, line 171 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qmwh6q7w7OHk2wNW:-LA-qmwh6q7w7OHk2wNX:bjtb9xf) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L171)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> org_list = cls._sys_admin_client.get_org_list()
>         for org in [o for o in org_list.Org if hasattr(org_list, 'Org')]:
>             if org.get('name').lower() == org_name.lower():
>                 cls._org_href = org.get('href')
> ```
</details>

What is the need for this logic again, after successful creation of Org?

---

*[neo_tests/environment.py, line 197 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-sXav2knuhCbCBU_E:-LA-sXav2knuhCbCBU_F:b-6tpqwu) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L197)):*
> ```Python
> 
>     @classmethod
>     def create_ovdc(cls):
> ```

Method names are little misleading. Either rename the methods to reflect the right operation i.e., create_ovdc_if_not_present() (or) separate the logic into getExistingVdc() and createOvdc()

(or) on a second thought after looking the entire file, separation of logic doesn't seem to be very necessary. If you decide to leave as-is, ensure you add comments and docstring at right places to convey the actual logic.

---

*[neo_tests/environment.py, line 231 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-sO2Q8dzDDdduY-95:-LA-sO2Q8dzDDdduY-96:b-2bo5m2) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L231)):*
<details><summary><i alt="summary">Quoted 12 lines of code…</i></summary>

> ```
> netpools = system.list_network_pools()
>         netpool_to_use = None
>         netpool_name = cls._config['vcd']['default_netpool_name']
>         if netpool_name is not '*':
>             for item in netpools:
>                 if item.get('name') == netpool_name:
>                     netpool_to_use = item.get('name')
>                     break
>  
>         if netpool_to_use is None:
>             print('Using first netpool in system viz. ' +
>                   netpools[0].get('name'))
> ```
</details>

I suggest to move  the netpools logic to separate method. This method is too long

---

*[neo_tests/environment.py, line 239 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-s80C-SjyW9ieOiKc:-LA-s80C-SjyW9ieOiKd:byw2ke) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L239)):*
> ```Python
>         vdc_resource = org.create_org_vdc(
>             ovdc_name,
>             cls._pvdc_name,
> ```

What if Pvdc is not present? throw an exception

---

*[neo_tests/environment.py, line 253 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-t1Sb-qvziNv9H973:-LA-t1Sb-qvziNv9H974:b-vm4swq) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L253)):*
> ```
> for vdc in org.list_vdcs():
>             if vdc.get('name') == ovdc_name:
>                 cls._ovdc_href = vdc.get('href')
> ```
Again, what is the need for this logic?

---

*[neo_tests/environment.py, line 257 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-tDeYDg-4cXixYiR6:-LA-tDeYDg-4cXixYiR7:bu4fst) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L257)):*
> ```Python
> 
>     @classmethod
>     def create_catalog(cls):
> ```

Same Rename method or separate logic to getExistingCatalog and createCatalog

---

*[neo_tests/environment.py, line 257 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vMtj2_BOr8DhfSKt:-LA-vMtj2_BOr8DhfSKu:bc9b7s8) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L257)):*
> ```Python
> 
>     @classmethod
>     def create_catalog(cls):
> ```

use the try block and put the client.logout() in the finally block.

---

*[neo_tests/environment.py, line 277 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-wQZY70UmkeQftdAL:-LA-wQZY70UmkeQftdAM:b-jko82j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L277)):*
> ```Python
>         catalog_resource = org.create_catalog(name=catalog_name,
>                                               description='')
>         catalog_author_client.get_task_monitor().wait_for_success(
> ```

Why are we not storing catalog href in cls object?

---

*[neo_tests/environment.py, line 300 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vF5S1V72Ag0OUGwO:-LA-vF5S1V72Ag0OUGwP:b6m84py) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L300)):*
> ```Python
> 
>     @classmethod
>     def upload_template(cls):
> ```

can you use the try block and put the client.logout() in finally?

---

*[neo_tests/environment.py, line 337 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vfNEBF6rGG9GHTtE:-LA-v5WG38Arf9cLGOUX:b-w7higc) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L337)):*
> ```Python
>             raise Exception('OVDC ' + cls._config['vcd']['default_ovdc_name'] +
>                             ' doesn\'t exist.')
>         catalog_author_client = Environment.get_client(
> ```

You mean vApp Author?

---

*[neo_tests/environment.py, line 344 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vn-j1EFwhgeGi8Yd:-LA-vn-j1EFwhgeGi8Ye:bwooxsr) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L344)):*
> ```
>    except Exception as e:
> ```
Modify this code to reflect right exception (ResourceNotFoundException) after Anusuya's change is checked-in or let her know of your changes.

---

*[neo_tests/environment.py, line 356 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-wXElE_CA6zdoS_ln:-LA-wXElE_CA6zdoS_lo:bi9yjyz) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L356)):*
> ```Python
>         finally:
>             catalog_author_client.logout()
>         print('Reusing existing vApp ' + vapp_name + '.')
> ```

why are we not using vApp href in cls object? any reason? 
Move the print statement to try block.

---

*[neo_tests/environment.py, line 388 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-xNbt5aOM4FUJaHM-:-LA-xNbt5aOM4FUJaHM0:b-uyxveg) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L388)):*
> ```Python
>         vapp_resource = vdc.get_vapp(vapp_name)
>         return VApp(client, resource=vapp_resource)
> ```

No method for get_default_catalog()?

---

*[neo_tests/neo_vcd_disk.py, line 3 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0-gYdAnmVZoozNEdC:-LA0-gYdAnmVZoozNEdD:b-c25gn8) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L3)):*
> ```Python
> import unittest
> 
> from neo_tests.base_test import BaseTestCase
> ```

Can we change neo_tests to system_tests (or) something that sounds similar?

---

*[neo_tests/neo_vcd_disk.py, line 11 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA01kTK2X6YMmoSu8aA:-LA01kTK2X6YMmoSu8aB:b-d7mdb6) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L11)):*
> ```Python
> 
> 
> class TestDisk(BaseTestCase):
> ```

I suggest get this test reviewed by Sompa, as she contributed to disk tests.
@sompa 

---

*[neo_tests/neo_vcd_disk.py, line 116 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA00ZIvAdugpFzDX0dN:-LA00ZIvAdugpFzDX0dO:b2pbi20) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L116)):*
> ```Python
>         new_owner_name = disk_resource.Owner.User.get('name')
> 
>         org_admin_client.logout()
> ```

Add try block ; move client logout into finally block
Do the same for other methods as well.

---

*[neo_tests/neo_vcd_disk.py, line 120 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02NCn1ksiipYyZBhC:-LA02NCn1ksiipYyZBhD:b-aitbj0) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L120)):*
> ```Python
>         self.assertEqual(new_owner_name, new_username)
> 
>     def test_0049_attach_detach_setup(self):
> ```

Shouldn't this be a private method invoked from test methods?

---

*[neo_tests/neo_vcd_disk.py, line 128 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02h_t-Q97NZkL-Of_:-LA02h_t-Q97NZkL-Ofa:b-8g5okg) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L128)):*
> ```Python
>         # TODO : update power_off to handle missing link exception
>         try:
>             task = vapp.power_off()
> ```

vapp must be back powered on after attach/detach test methods.
Any operation done as part of setup must be reverted.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
I see a lot of print statements.  We instead should use a logger and send output to a file combined with pyvcloud output so that it's easier to associate test activities with pyvcloud.  Also, using print() makes it hard to read the unittest output. 

---

Review status: all files reviewed at latest revision, 28 unresolved discussions.

---

*[neo_tests/base_config.yaml, line 1 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0Godh540PxvNNeeCG:-LA0Godh540PxvNNeeCH:bl5ux9e) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L1)):*
> ```YAML
> global:
> ```

Please add comments to this file so it's clear what values are used for.  We need to make this easy for new devs to fill it out.  (See examples/tenant.yaml for suggested style--more information is better.)

Also, it would be good go distinguish in the comments which values *must* be changed  vs. those that normally don't need to be--example of the latter seems to be most of the values starting with 'default'.  

---

*[neo_tests/base_config.yaml, line 4 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-6SPr87B_H2my-v2D:-LA0GJFn51wyG5O5rzr2:b2cmk9h) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L4)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

These configs are related to the rest calls we are making.  Right now it's not working as designed, since we are still getting warnings, I can rename t to something else. Any suggestions?
</blockquote></details>

How about 'connection'?

---

*[neo_tests/base_test.py, line 6 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0HszeE-jY2TFapSzc:-LA0HszeE-jY2TFapSzd:b-37s3lz) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L6)):*
> ```Python
> import yaml
> 
> from neo_tests.environment import Environment
> ```

This class and Environment should live in pyvcloud/vcd so that the test framework can be reused in vcd-cli. 

---

*[neo_tests/neo_vcd_disk.py, line 11 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0K84MDnKCJkxAE0uc:-LA0K84MDnKCJkxAE0ud:b-ggi8w9) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L11)):*
> ```Python
> 
> 
> class TestDisk(BaseTestCase):
> ```

Please add a docstring to describe the overall purpose of this test. 

---

*[neo_tests/neo_vcd_disk.py, line 73 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0J_xs9XpjpEY8eGtu:-LA0J_xs9XpjpEY8eGtv:b-k40rlk) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L73)):*
> ```Python
>         return disk_id
> 
>     def test_0010_get_all_disks(self):
> ```

Please add a docstring for each case that tells what it tests.  python -m unittest will print these which makes test output easy to read and diagnose quickly. 

---

*[neo_tests/neo_vcd_disk.py, line 73 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0KLMQ3PGzpzgHBrsi:-LA0KLMQ3PGzpzgHBrsj:b-x6un87) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L73)):*
> ```Python
>         return disk_id
> 
>     def test_0010_get_all_disks(self):
> ```

Separate question--are the test cases independent?  If not what happens if one of them fails--do the remaining cases fail as well?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-:-LA0ITvGDNyZTMTzcXD-:bxvvux2)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Review status: all files reviewed at latest revision, 33 unresolved discussions.

---

*[neo_tests/base_test.py, line 6 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0HszeE-jY2TFapSzc:-LA0XCy157oi2DVvPh3e:b-t37t3y) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L6)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

This class and Environment should live in pyvcloud/vcd so that the test framework can be reused in vcd-cli. 
</blockquote></details>

We install the entire pyvcloud project in vcd-cli, the test modules should be accessible from vcd-cli,
However I think we should move the whole testing folder inside the second pyvcloud folder (lateral to vcd folder)

So that out imports in vcd-cli looks like
_from pyvcloud.system_tests import Environment_

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 32 unresolved discussions.

---

*[a discussion](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0bBee2PBcsqWiaKIT:-LA0bBee2PBcsqWiaKIU:b-gltk9b) (no related file):*
Overall I like the structure of the new test.  This form looks as if it will be relatively easy to imitate in the remaining tests and it's not far from how we do things already.   Ready to sign off as soon as other comments regarding logging and comments are resolved. 

---

*[neo_tests/base_test.py, line 6 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0HszeE-jY2TFapSzc:-LA0c3VS7unIL26cewvb:b-wxf1qk) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L6)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We install the entire pyvcloud project in vcd-cli, the test modules should be accessible from vcd-cli,
However I think we should move the whole testing folder inside the second pyvcloud folder (lateral to vcd folder)

So that out imports in vcd-cli looks like
_from pyvcloud.system_tests import Environment_


</blockquote></details>

I would prefer to keep the current tests directory for actual tests and just put infra classes that other packages may need in pyvcloud.  Tests as it turns out are not accessible as there's no module name for them which is good--keeps the namespace from getting cluttered.  We also don't want them in the Sphinx docs.  Just Environment and BaseTestCase should go in, much as we do now. 

---

*[neo_tests/environment.py, line 162 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qHTfA5SZDwr0geXF:-LA0XVR79xGvyyRlxHMa:bvebkyj) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L162)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Sure, we can do that. But, I will like to loop Robert in before taking a decission to pipe the output to log files away from stdout.
</blockquote></details>

I"m also in favor of logging as noted in top comment.  Otherwise the test output is extremely hard to read in builds as it's cluttered with extra info that's only interesting if there's a failure. 

---

*[neo_tests/neo_vcd_disk.py, line 36 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0_O9w72T7GkLVo_RU:-LA0_O9w72T7GkLVo_RV:b-67ro2l) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L36)):*
> ```Python
>         vdc = Environment.get_default_vdc(TestDisk._client)
> 
>         TestDisk._idisk1_id = self._create_disk_helper(
> ```

Will this fail if a disk already exists with that name?  There's no guarantee cleanup will run if a previous test fails. 

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

sahithi commented:
Review status: all files reviewed at latest revision, 28 unresolved discussions.

---

*[neo_tests/environment.py, line 337 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vfNEBF6rGG9GHTtE:-LAF0ZTZ-2Wzd1act0Xv:b-8wknb2) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L337)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Currently we have a bug in pyvcloud, when we share a catalog we don't share it with all users  but we share it will all org-admins of other organizations. So vApp author won't even see the catalog :( Hence all my tests are written as catalog author.

To reproduce the issue in vcd-cli  create a catalog as catalog author, add something to it, and thn try to list the content of the catalog. You will get an error : 'catalog not found'
</blockquote></details>

Please mention the same here as TODO and do add this item to Umbrella Jira task

---

*[neo_tests/environment.py, line 356 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-wXElE_CA6zdoS_ln:-LAF0rs9ELv25iKzgu2n:b5sg7m1) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L356)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We can retieve vApps by name, so didn't see a need to cache the href.
</blockquote></details>

Retrieving vApp by name is an expensive call (get_vapp() is scanning all vApps in a given vdc).

---

*[neo_tests/neo_vcd_disk.py, line 120 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02NCn1ksiipYyZBhC:-LAF2F0R9b2N6sQ0pbyI:b-fs5tfr) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L120)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We can leave it as a public test method, as unit test will run this before the actual attach method. We can make it a private function and call it explicitly from the attach test too. I don't see any benefit in doing things via one way over other.
</blockquote></details>

Any method marked public indicates that it is the actual test Not helper methods, rest everything should be marked private.
I am a bit particular about this because our coding style in first few check-ins of the tests will set the standard for both internal and external contributors.

---

*[neo_tests/neo_vcd_disk.py, line 128 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02h_t-Q97NZkL-Of_:-LAF36jk7yCzUvYOW3GU:bam6gjp) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L128)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

This one will need a little bit of more work. I will probably need to implement method in vApp.py to get the power state of a vm/vApp. 
Currently I don't know what the state of vm is so had to put the whole thing in a try catch block. To restore the state I first need to know the state the vm was in before the test.
</blockquote></details>

You don't have to implement it now. But do add a TODO here.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Added JIRA VCDA-604 to change print() statements to logger statements.

---

Reviewed 4 of 9 files at r1, 2 of 3 files at r2.
Review status: 0 of 5 files reviewed at latest revision, 28 unresolved discussions.

---

*[neo_tests/base_config.yaml, line 1 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0Godh540PxvNNeeCG:-LAH6OssEZE-69MMZo4K:bb6un1j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L1)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Please add comments to this file so it's clear what values are used for.  We need to make this easy for new devs to fill it out.  (See examples/tenant.yaml for suggested style--more information is better.)

Also, it would be good go distinguish in the comments which values *must* be changed  vs. those that normally don't need to be--example of the latter seems to be most of the values starting with 'default'.  
</blockquote></details>

Done

---

*[neo_tests/base_config.yaml, line 4 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-6SPr87B_H2my-v2D:-LAGtlRb4IUf5qN-63Kv:b-8o2if0) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_config.yaml#L4)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

How about 'connection'?
</blockquote></details>

+1 :)

---

*[neo_tests/base_test.py, line 6 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0HszeE-jY2TFapSzc:-LAH6QZ7EOP13KCIOVpv:bw9vt5l) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/base_test.py#L6)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

I would prefer to keep the current tests directory for actual tests and just put infra classes that other packages may need in pyvcloud.  Tests as it turns out are not accessible as there's no module name for them which is good--keeps the namespace from getting cluttered.  We also don't want them in the Sphinx docs.  Just Environment and BaseTestCase should go in, much as we do now. 
</blockquote></details>

Moved around the framework code in pyvcloud/pyvcloud/system_test_framework and tests into pyvcloud/system_tests

---

*[neo_tests/environment.py, line 2 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-oIU_5Ci4nYeKv5ay:-LAGu5dQ08gWZrdvWqHr:b-8b11j9) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L2)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Add docstrings for all the methods.
</blockquote></details>

Added

---

*[neo_tests/environment.py, line 120 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-ncVQ4W1dz_UAq1az:-LAGuB287RKA2JZrmOtr:b4rfsh0) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L120)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

> ```
> # Untested code
>         platform.attach_vcenter(
>             vc_server_name=cls._config['vc']['vcenter_host_name'],
>             vc_server_host=cls._config['vc']['vcenter_host_ip'],
>             vc_admin_user=cls._config['vc']['vcenter_admin_username'],
>             vc_admin_pwd=cls._config['vc']['vcenter_admin_password'],
>             nsx_server_name=cls._config['nsx']['nsx_hostname'],
>             nsx_host=cls._config['nsx']['nsx_host_ip'],
>             nsx_admin_user=cls._config['nsx']['nsx_admin_username'],
>             nsx_admin_pwd=cls._config['nsx']['nsx_admin_password'],
>             is_enabled=True)
> ```

Please file a Jira task to track this and mention the taskId in TODO here
</blockquote></details>

Added JIRA task VCDA-603

---

*[neo_tests/environment.py, line 137 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-p9FY1vmH4VTI24T6:-LAGuDj-6ARxGSE1CXn2:b-886gsx) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L137)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

nit: reusing
</blockquote></details>

Fixed

---

*[neo_tests/environment.py, line 160 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qTDz-EK96lcUVMgZ:-LAGuFO6AqWqOSlrDtem:bh78x9n) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L160)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

consider clubbing above two print statements
</blockquote></details>

Removed the later print statement as it was for my own testing.

---

*[neo_tests/environment.py, line 162 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-qHTfA5SZDwr0geXF:-LAGFKiw7C2F2Y8aLlIO:b-uu59ty) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L162)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

I"m also in favor of logging as noted in top comment.  Otherwise the test output is extremely hard to read in builds as it's cluttered with extra info that's only interesting if there's a failure. 
</blockquote></details>

I have filed a new JIRA task to convert all print statements to loglines. I haven't used the logger in pyvcloud so I will need some time to read up on how it's being used in pyvcloud. Don't want to stop this current CL from being checked in while I am working on the logger.

VCDA-604

---

*[neo_tests/environment.py, line 231 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-sO2Q8dzDDdduY-95:-LAGuM12-W84YcEEN7vx:b-o3skf8) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L231)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

> ```
> netpools = system.list_network_pools()
>         netpool_to_use = None
>         netpool_name = cls._config['vcd']['default_netpool_name']
>         if netpool_name is not '*':
>             for item in netpools:
>                 if item.get('name') == netpool_name:
>                     netpool_to_use = item.get('name')
>                     break
>  
>         if netpool_to_use is None:
>             print('Using first netpool in system viz. ' +
>                   netpools[0].get('name'))
> ```
I suggest to move  the netpools logic to separate method. This method is too long
</blockquote></details>

Moved. On a side note, compared to the instantiate_vapp method in vdc.py everything is a toddler.

---

*[neo_tests/environment.py, line 239 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-s80C-SjyW9ieOiKc:-LAGuUoZ9Iure9-vPqGb:bxsogr8) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L239)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

What if Pvdc is not present? throw an exception
</blockquote></details>

Added code to do so.

---

*[neo_tests/environment.py, line 253 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-t1Sb-qvziNv9H973:-LAGGRBoBL9zVL6b8p8H:b-7yrrco) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L253)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

> ```
> for vdc in org.list_vdcs():
>             if vdc.get('name') == ovdc_name:
>                 cls._ovdc_href = vdc.get('href')
> ```
Again, what is the need for this logic?
</blockquote></details>

We create the ovdc as sys admin, the href that is returned as part of the create_org_vdc() call is a admin only view i.e. it's like /api/admin/vdc/..
We want the non admin href  hence the logic.

---

*[neo_tests/environment.py, line 257 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-tDeYDg-4cXixYiR6:-LAGGsEL1ZnidkqcSOtW:b-ypsy4) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L257)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Same Rename method or separate logic to getExistingCatalog and createCatalog
</blockquote></details>

I think this is not necessary, all these methods follow the same pattern : grab the resource if present else create it. These methods are not for consumption by end user.

---

*[neo_tests/environment.py, line 257 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vMtj2_BOr8DhfSKt:-LAGuYt-9MUNOSgDhKVb:bb6un1j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L257)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

use the try block and put the client.logout() in the finally block.
</blockquote></details>

Done

---

*[neo_tests/environment.py, line 300 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vF5S1V72Ag0OUGwO:-LAGuauD-pJ6iFUKFCYu:bb6un1j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L300)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

can you use the try block and put the client.logout() in finally?
</blockquote></details>

Done

---

*[neo_tests/environment.py, line 337 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vfNEBF6rGG9GHTtE:-LAGucwTEsgWjbDAj52x:b4rfsh0) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L337)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Please mention the same here as TODO and do add this item to Umbrella Jira task
</blockquote></details>

Added JIRA task VCDA-603

---

*[neo_tests/environment.py, line 344 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-vn-j1EFwhgeGi8Yd:-LAGLzvq1e8l8VcPXQvd:bk9j7pr) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L344)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

> ```
>    except Exception as e:
> ```
Modify this code to reflect right exception (ResourceNotFoundException) after Anusuya's change is checked-in or let her know of your changes.
</blockquote></details>

Adding this to the umbrella JIRA task VCDA-603

---

*[neo_tests/environment.py, line 356 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-wXElE_CA6zdoS_ln:-LAGj-503xQg1KMka-bd:b-eekc6g) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L356)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Retrieving vApp by name is an expensive call (get_vapp() is scanning all vApps in a given vdc).
</blockquote></details>

Agreed. Adding the required field

---

*[neo_tests/neo_vcd_disk.py, line 3 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0-gYdAnmVZoozNEdC:-LAGj34L0bm793nly-4s:b-896fix) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L3)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Can we change neo_tests to system_tests (or) something that sounds similar?
</blockquote></details>

Done.

---

*[neo_tests/neo_vcd_disk.py, line 11 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA01kTK2X6YMmoSu8aA:-LAGuluw22F70tUepQKv:bm8lr4f) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L11)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

I suggest get this test reviewed by Sompa, as she contributed to disk tests.
@sompa 
</blockquote></details>

Right now I have not ported iops tests. Testing iops requires extra work on VC storage profile which we can't do from within pyvcloud. I think we will need a separate discussion on how to make sure VC is configured in the right way before our tests are run.

On a separate note, I don't think casini testbeds have iops configured storage profiles.

---

*[neo_tests/neo_vcd_disk.py, line 11 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0K84MDnKCJkxAE0uc:-LAH4SoL5_9F7yPqEoae:b-8b11j9) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L11)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Please add a docstring to describe the overall purpose of this test. 
</blockquote></details>

Added

---

*[neo_tests/neo_vcd_disk.py, line 36 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0_O9w72T7GkLVo_RU:-LA0nVWe5Nheon_tIOLx:b-dtfyg9) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L36)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Will this fail if a disk already exists with that name?  There's no guarantee cleanup will run if a previous test fails. 
</blockquote></details>

No, it will not fail. There can be multiple independent disk with the same name.
Btw if a test fails. Won't we reset the whole testbed back to it's original state? (We were discussing about this yesterday and ways to do it, snapshot being one).

---

*[neo_tests/neo_vcd_disk.py, line 73 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0J_xs9XpjpEY8eGtu:-LAH4yRFAg8nQEVc11d6:b-fqamm3) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L73)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

Please add a docstring for each case that tells what it tests.  python -m unittest will print these which makes test output easy to read and diagnose quickly. 
</blockquote></details>

I searched around a bit and it seems we don't add docstrings to unittest tests, rather the test function name should be descriptive enough. Also we should run the tests with -v option for more verbose outputs.

---

*[neo_tests/neo_vcd_disk.py, line 116 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA00ZIvAdugpFzDX0dN:-LAH6vA8-F4nafAb6h3g:bb6un1j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L116)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Add try block ; move client logout into finally block
Do the same for other methods as well.
</blockquote></details>

Done

---

*[neo_tests/neo_vcd_disk.py, line 120 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02NCn1ksiipYyZBhC:-LAH5SXf3twNJ5xdv0Mm:bsrndmf) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L120)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

Any method marked public indicates that it is the actual test Not helper methods, rest everything should be marked private.
I am a bit particular about this because our coding style in first few check-ins of the tests will set the standard for both internal and external contributors.

</blockquote></details>

Made the required changes

---

*[neo_tests/neo_vcd_disk.py, line 128 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA02h_t-Q97NZkL-Of_:-LAGwnZw91q8xq_4jIY4:bb6un1j) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L128)):*
<details><summary><i>Previously, sahithi (Sahithi Ayloo) wrote…</i></summary><blockquote>

You don't have to implement it now. But do add a TODO here.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-:-LAGt_l1Bpjih26GHrI2:b1w68vw)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Review status: 0 of 5 files reviewed at latest revision, 27 unresolved discussions.

---

*[neo_tests/environment.py, line 253 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-t1Sb-qvziNv9H973:-LAJ1bhdB3VJhSzEzfRF:b32w0x) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L253)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

We create the ovdc as sys admin, the href that is returned as part of the create_org_vdc() call is a admin only view i.e. it's like /api/admin/vdc/..
We want the non admin href  hence the logic.
</blockquote></details>

OK

---

*[neo_tests/neo_vcd_disk.py, line 36 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0_O9w72T7GkLVo_RU:-LAJ1-4o4efpsQK4OOMJ:b-7piw9r) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L36)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

No, it will not fail. There can be multiple independent disk with the same name.
Btw if a test fails. Won't we reset the whole testbed back to it's original state? (We were discussing about this yesterday and ways to do it, snapshot being one).
</blockquote></details>

It's better if the test does it.  Snapshots are slow and complicated; they won't work on a shared vCD.  Also, imagine you are a dev just rerunning it until something works--it's way faster if setup just guarantees the preconditions necessary to run the test cases.

---

*[neo_tests/neo_vcd_disk.py, line 73 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0J_xs9XpjpEY8eGtu:-LAJ0J4O9RPEJJJGNALD:bsm5471) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L73)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I searched around a bit and it seems we don't add docstrings to unittest tests, rather the test function name should be descriptive enough. Also we should run the tests with -v option for more verbose outputs.
</blockquote></details>

I would still advocate doing this.  Function names are pretty short.  Also, it's a good discipline to think of each test verifying a particular logical proposition.  This used to be called 'assertion-based testing'; it's a very good way of designing specification-based tests.  You can pretty much read out the coverage from the tests.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

rocknes commented:
Review status: 0 of 5 files reviewed at latest revision, 25 unresolved discussions.

---

*[neo_tests/neo_vcd_disk.py, line 36 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0_O9w72T7GkLVo_RU:-LAJqgKu1y-qQXIjC42d:b-wbmcrf) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L36)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

It's better if the test does it.  Snapshots are slow and complicated; they won't work on a shared vCD.  Also, imagine you are a dev just rerunning it until something works--it's way faster if setup just guarantees the preconditions necessary to run the test cases.
</blockquote></details>

I am a little unclear about your suggestion here. Are you suggesting that before the setup creates the disk it should check for existence of the disk? And if those disk exist we will just re-use them? If that's the ask, I will add the required logic, shouldn't be too hard to do it :)

---

*[neo_tests/neo_vcd_disk.py, line 73 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0J_xs9XpjpEY8eGtu:-LAJqytk8BUElFrn-IdZ:bwmi3bh) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L73)):*
<details><summary><i>Previously, hodgesrm (Robert Hodges) wrote…</i></summary><blockquote>

I would still advocate doing this.  Function names are pretty short.  Also, it's a good discipline to think of each test verifying a particular logical proposition.  This used to be called 'assertion-based testing'; it's a very good way of designing specification-based tests.  You can pretty much read out the coverage from the tests.
</blockquote></details>

Sure, I will add the docstrings.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Review status: 0 of 5 files reviewed at latest revision, 25 unresolved discussions.

---

*[neo_tests/neo_vcd_disk.py, line 36 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA0_O9w72T7GkLVo_RU:-LAJvK9w5ykRpy2JQsR5:b-tjtdmc) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/neo_vcd_disk.py#L36)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

I am a little unclear about your suggestion here. Are you suggesting that before the setup creates the disk it should check for existence of the disk? And if those disk exist we will just re-use them? If that's the ask, I will add the required logic, shouldn't be too hard to do it :)
</blockquote></details>

Exactly.  That would probably do it for now unless you are counting on the state of the bits your tests.  Thanks!

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Reviewed 2 of 3 files at r2.
Review status: 0 of 5 files reviewed at latest revision, 25 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thank you Aritra.  This puppy looks ready to go.  

---

Reviewed 1 of 3 files at r3.
Review status: 0 of 5 files reviewed at latest revision, 24 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-:-LAL0bn4BjYrmRwYshX9:bvwgpgd)*
<!-- Sent from Reviewable.io -->

hodgesrm commented:
Reviewed 1 of 3 files at r3.
Review status: 0 of 5 files reviewed at latest revision, 23 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227)*
<!-- Sent from Reviewable.io -->

sahithi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r2, 2 of 3 files at r3.
Review status: 0 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[neo_tests/environment.py, line 231 at r1](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-LA-sO2Q8dzDDdduY-95:-LAQ8BNb5WLPLP8gH7lt:benykj1) ([raw file](https://github.com/vmware/pyvcloud/blob/aa7d61c7a09df3a839dd891bc9819c924985638b/neo_tests/environment.py#L231)):*
<details><summary><i>Previously, rocknes wrote…</i></summary><blockquote>

Moved. On a side note, compared to the instantiate_vapp method in vdc.py everything is a toddler.
</blockquote></details>

Don't underestimate toddlers Aritra.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vmware/pyvcloud/227#-:-LAQAPNY2JcveguhORwD:bnfp4nl)*
<!-- Sent from Reviewable.io -->
